man/man8/verify.8
.TH VERIFY 8 
.ad
.fi
.SH NAME
verify
\-
Postfix address verification server
.SH "SYNOPSIS"
.na
.nf
\fBverify\fR [generic Postfix daemon options]
.SH DESCRIPTION
.ad
.fi
The \fBverify\fR(8) address verification server maintains a record
of what recipient addresses are known to be deliverable or
undeliverable.

Addresses are verified by injecting probe messages into the
Postfix queue. Probe messages are run through all the routing
and rewriting machinery except for final delivery, and are
discarded rather than being deferred or bounced.

Address verification relies on the answer from the nearest
MTA for the specified address, and will therefore not detect
all undeliverable addresses.

The \fBverify\fR(8) server is designed to run under control
by the Postfix
master server. It maintains an optional persistent database.
To avoid being interrupted by "postfix stop" in the middle
of a database update, the process runs in a separate process
group.

The \fBverify\fR(8) server implements the following requests:
.IP "\fBupdate\fI address status text\fR"
Update the status and text of the specified address.
.IP "\fBquery\fI address\fR"
Look up the \fIstatus\fR and \fItext\fR for the specified
\fIaddress\fR.
If the status is unknown, a probe is sent and an "in progress"
status is returned.
.SH "SECURITY"
.na
.nf
.ad
.fi
The address verification server is not security\-sensitive. It does
not talk to the network, and it does not talk to local users.
The verify server can run chrooted at fixed low privilege.

The address verification server can be coerced to store
unlimited amounts of garbage. Limiting the cache expiry
time
trades one problem (disk space exhaustion) for another
one (poor response time to client requests).

With Postfix version 2.5 and later, the \fBverify\fR(8)
server no longer uses root privileges when opening the
\fBaddress_verify_map\fR cache file. The file should now
be stored under the Postfix\-owned \fBdata_directory\fR.  As
a migration aid, an attempt to open a cache file under a
non\-Postfix directory is redirected to the Postfix\-owned
\fBdata_directory\fR, and a warning is logged.
.SH DIAGNOSTICS
.ad
.fi
Problems and transactions are logged to \fBsyslogd\fR(8)
or \fBpostlogd\fR(8).
.SH BUGS
.ad
.fi
Address verification probe messages add additional traffic
to the mail queue.
Recipient verification may cause an increased load on
down\-stream servers in the case of a dictionary attack or
a flood of backscatter bounces.
Sender address verification may cause your site to be
denylisted by some providers.

If the persistent database ever gets corrupted then the world
comes to an end and human intervention is needed. This violates
a basic Postfix principle.
.SH "CONFIGURATION PARAMETERS"
.na
.nf
.ad
.fi
Changes to \fBmain.cf\fR are not picked up automatically,
as \fBverify\fR(8)
processes are long\-lived. Use the command "\fBpostfix reload\fR" after
a configuration change.

The text below provides only a parameter summary. See
\fBpostconf\fR(5) for more details including examples.
.SH "PROBE MESSAGE CONTROLS"
.na
.nf
.ad
.fi
.IP "\fBaddress_verify_sender ($double_bounce_sender)\fR"
The sender address to use in address verification probes; prior
to Postfix 2.5 the default was "postmaster".
.PP
Available with Postfix 2.9 and later:
.IP "\fBaddress_verify_sender_ttl (0s)\fR"
The time between changes in the time\-dependent portion of address
verification probe sender addresses.
.SH "CACHE CONTROLS"
.na
.nf
.ad
.fi
.IP "\fBaddress_verify_map (see 'postconf -d' output)\fR"
Lookup table for persistent address verification status
storage.
.IP "\fBaddress_verify_positive_expire_time (31d)\fR"
The time after which a successful probe expires from the address
verification cache.
.IP "\fBaddress_verify_positive_refresh_time (7d)\fR"
The time after which a successful address verification probe needs
to be refreshed.
.IP "\fBaddress_verify_negative_cache (yes)\fR"
Enable caching of failed address verification probe results.
.IP "\fBaddress_verify_negative_expire_time (3d)\fR"
The time after which a failed probe expires from the address
verification cache.
.IP "\fBaddress_verify_negative_refresh_time (3h)\fR"
The time after which a failed address verification probe needs to
be refreshed.
.PP
Available with Postfix 2.7 and later:
.IP "\fBaddress_verify_cache_cleanup_interval (12h)\fR"
The amount of time between \fBverify\fR(8) address verification
database cleanup runs.
.SH "PROBE MESSAGE ROUTING CONTROLS"
.na
.nf
.ad
.fi
By default, probe messages are delivered via the same route
as regular messages.  The following parameters can be used to
override specific message routing mechanisms.
.IP "\fBaddress_verify_relayhost ($relayhost)\fR"
Overrides the relayhost parameter setting for address verification
probes.
.IP "\fBaddress_verify_transport_maps ($transport_maps)\fR"
Overrides the transport_maps parameter setting for address verification
probes.
.IP "\fBaddress_verify_local_transport ($local_transport)\fR"
Overrides the local_transport parameter setting for address
verification probes.
.IP "\fBaddress_verify_virtual_transport ($virtual_transport)\fR"
Overrides the virtual_transport parameter setting for address
verification probes.
.IP "\fBaddress_verify_relay_transport ($relay_transport)\fR"
Overrides the relay_transport parameter setting for address
verification probes.
.IP "\fBaddress_verify_default_transport ($default_transport)\fR"
Overrides the default_transport parameter setting for address
verification probes.
.PP
Available in Postfix 2.3 and later:
.IP "\fBaddress_verify_sender_dependent_relayhost_maps ($sender_dependent_relayhost_maps)\fR"
Overrides the sender_dependent_relayhost_maps parameter setting for address
verification probes.
.PP
Available in Postfix 2.7 and later:
.IP "\fBaddress_verify_sender_dependent_default_transport_maps ($sender_dependent_default_transport_maps)\fR"
Overrides the sender_dependent_default_transport_maps parameter
setting for address verification probes.
.SH "SMTPUTF8 CONTROLS"
.na
.nf
.ad
.fi
Preliminary SMTPUTF8 support is introduced with Postfix 3.0.
.IP "\fBsmtputf8_autodetect_classes (sendmail, verify)\fR"
Detect that a message requires SMTPUTF8 support for the specified
mail origin classes.
.PP
Available in Postfix version 3.2 and later:
.IP "\fBenable_idna2003_compatibility (no)\fR"
Enable 'transitional' compatibility between IDNA2003 and IDNA2008,
when converting UTF\-8 domain names to/from the ASCII form that is
used for DNS lookups.
.SH "MISCELLANEOUS CONTROLS"
.na
.nf
.ad
.fi
.IP "\fBconfig_directory (see 'postconf -d' output)\fR"
The default location of the Postfix main.cf and master.cf
configuration files.
.IP "\fBdaemon_timeout (18000s)\fR"
How much time a Postfix daemon process may take to handle a
request before it is terminated by a built\-in watchdog timer.
.IP "\fBipc_timeout (3600s)\fR"
The time limit for sending or receiving information over an internal
communication channel.
.IP "\fBprocess_id (read\-only)\fR"
The process ID of a Postfix command or daemon process.
.IP "\fBprocess_name (read\-only)\fR"
The process name of a Postfix command or daemon process.
.IP "\fBqueue_directory (see 'postconf -d' output)\fR"
The location of the Postfix top\-level queue directory.
.IP "\fBsyslog_facility (mail)\fR"
The syslog facility of Postfix logging.
.IP "\fBsyslog_name (see 'postconf -d' output)\fR"
A prefix that is prepended to the process name in syslog
records, so that, for example, "smtpd" becomes "prefix/smtpd".
.PP
Available in Postfix 3.3 and later:
.IP "\fBservice_name (read\-only)\fR"
The master.cf service name of a Postfix daemon process.
.SH "SEE ALSO"
.na
.nf
smtpd(8), Postfix SMTP server
cleanup(8), enqueue Postfix message
postconf(5), configuration parameters
postlogd(8), Postfix logging
syslogd(8), system logging
.SH "README FILES"
.na
.nf
.ad
.fi
Use "\fBpostconf readme_directory\fR" or
"\fBpostconf html_directory\fR" to locate this information.
.na
.nf
ADDRESS_VERIFICATION_README, address verification howto
.SH "LICENSE"
.na
.nf
.ad
.fi
The Secure Mailer license must be distributed with this software.
.SH HISTORY
.ad
.fi
.ad
.fi
This service was introduced with Postfix version 2.1.
.SH "AUTHOR(S)"
.na
.nf
Wietse Venema
IBM T.J. Watson Research
P.O. Box 704
Yorktown Heights, NY 10598, USA

Wietse Venema
Google, Inc.
111 8th Avenue
New York, NY 10011, USA


RELEASE_NOTES
This is the Postfix 3.11 experimental release.

The stable Postfix release is called postfix-3.10.x where 3=major
release number, 10=minor release number, x=patchlevel. The stable
release never changes except for patches that address bugs or
emergencies. Patches change the patchlevel and the release date.

New features are developed in snapshot releases. These are called
postfix-3.11-yyyymmdd where yyyymmdd is the release date (yyyy=year,
mm=month, dd=day). Patches are never issued for snapshot releases;
instead, a new snapshot is released.

The mail_release_date configuration parameter (format: yyyymmdd)
specifies the release date of a stable release or snapshot release.

If you upgrade from Postfix 3.9 or earlier, please read RELEASE_NOTES-3.10
before proceeding.

Dual license
------------

As of Postfix 3.2.5 this software is distributed with a dual license:
in addition to the historical IBM Public License (IPL) 1.0, it is
now also distributed with the more recent Eclipse Public License
(EPL) 2.0. Recipients can choose to take the software under the
license of their choice. Those who are more comfortable with the
IPL can continue with that license.


postfix/src/global/verify.c
/*++
/* NAME
/*	verify 3
/* SUMMARY
/*	update verify database
/* SYNOPSIS
/*	#include <verify.h>
/*
/*	int	verify_append(queue_id, stats, recipient, relay, tstats, dsn,
/*				verify_status)
/*	const char *queue_id;
/*	MSG_STATS *stats;
/*	RECIPIENT *recipient;
/*	const char *relay;
/*	const POL_STATS *tstats;
/*	DSN	*dsn;
/*	int	verify_status;
/* DESCRIPTION
/*	This module implements an impedance adaptor between the
/*	verify_clnt interface and the interface expected by the
/*	bounce/defer/sent modules.
/*
/*	verify_append() updates the address verification database
/*	and logs the action to the mailer logfile.
/*
/*	Arguments:
/* .IP queue_id
/*	The message queue id.
/* .IP stats
/*	Time stamps from different message delivery stages
/*	and session reuse count.
/* .IP recipient
/*	Recipient information. See recipient_list(3).
/* .IP relay
/*	Name of the host we're talking to.
/* .IP tstats
/*	TLS per-feature status.
/* .IP dsn
/*	Delivery status information. See dsn(3).
/*	The action is one of "deliverable" or "undeliverable".
/* .IP verify_status
/*	One of the following recipient verification status codes:
/* .RS
/* .IP DEL_REQ_RCPT_STAT_OK
/*	Successful delivery.
/* .IP DEL_REQ_RCPT_STAT_DEFER
/*	Temporary delivery error.
/* .IP DEL_REQ_RCPT_STAT_BOUNCE
/*	Hard delivery error.
/* .RE
/* DIAGNOSTICS
/*	A non-zero result means the operation failed.
/*
/*	Fatal: out of memory.
/* BUGS
/*	Should be replaced by routines with an attribute-value based
/*	interface instead of an interface that uses a rigid argument list.
/* LICENSE
/* .ad
/* .fi
/*	The Secure Mailer license must be distributed with this software.
/* AUTHOR(S)
/*	Wietse Venema
/*	IBM T.J. Watson Research
/*	P.O. Box 704
/*	Yorktown Heights, NY 10598, USA
/*
/*	Wietse Venema
/*	Google, Inc.
/*	111 8th Avenue
/*	New York, NY 10011, USA
/*
/*	Wietse Venema
/*	porcupine.org
/*--*/

/* System library. */

#include <sys_defs.h>
#include <string.h>

/* Utility library. */

#include <msg.h>
#include <vstring.h>
#include <stringops.h>

/* Global library. */

#include <mail_params.h>
#include <mail_proto.h>
#include <verify_clnt.h>
#include <log_adhoc.h>
#include <verify.h>

/* verify_append - update address verification database */

int     verify_append(const char *queue_id, MSG_STATS *stats,
		              RECIPIENT *recipient, const char *relay,
		              const POL_STATS *tstats, DSN *dsn,
		              int vrfy_stat)
{
    int     req_stat;
    DSN     my_dsn = *dsn;

    /*
     * Impedance adaptor between bounce/defer/sent and verify_clnt.
     * 
     * XXX No DSN check; this routine is called from bounce/defer/sent, which
     * know what the DSN initial digit should look like.
     * 
     * XXX vrfy_stat is completely redundant because of dsn.
     */
    if (var_verify_neg_cache || vrfy_stat == DEL_RCPT_STAT_OK) {
	if (recipient->orig_addr[0])
	    req_stat = verify_clnt_update(recipient->orig_addr, vrfy_stat,
					  my_dsn.reason);
	else
	    req_stat = VRFY_STAT_OK;
	/* Two verify updates for one verify request! */
	if (req_stat == VRFY_STAT_OK
	    && strcmp(recipient->address, recipient->orig_addr) != 0)
	    req_stat = verify_clnt_update(recipient->address, vrfy_stat,
					  my_dsn.reason);
    } else {
	my_dsn.action = "undeliverable-but-not-cached";
	req_stat = VRFY_STAT_OK;
    }
    if (req_stat == VRFY_STAT_OK) {
	log_adhoc(queue_id, stats, recipient, relay, tstats, dsn,
		  my_dsn.action);
	req_stat = 0;
    } else {
	msg_warn("%s: %s service failure", queue_id, var_verify_service);
	req_stat = -1;
    }
    return (req_stat);
}
